code/utils_model.py

```diff
@@ -1,479 +1,509 @@
-import copy
-import time
-import random
-from sklearn.metrics import confusion_matrix
-import operator
+from PIL import Image
+from torch.optim import lr_scheduler
+from torchvision import (datasets, transforms)
+
+import config
+from utils import (get_image_paths, get_subfolder_paths)
 
 ###########################################
 #             MISC FUNCTIONS              #
 ###########################################
 
-PATH_MEAN = [0.7, 0.6, 0.7]
-PATH_STD = [0.15, 0.15, 0.15]
-
-
-def reverse_norm(mean, std):
-    mean = torch.tensor(mean, dtype=torch.float32)
-    std = torch.tensor(std, dtype=torch.float32)
-    return (-mean / std).tolist(), (1 / std).tolist()
-
-
-# computing the confusion matrix
-def update_confusion_matrix(all_labels, all_predicts, batch_labels,
-                            batch_predicts, num_classes):
-    if all_labels is not None:
-        assert all_labels.shape[0] == all_predicts.shape[0]
-        if all_labels.shape[0] > 10000:
-            all_labels = all_labels[-10000:]
-            all_predicts = all_predicts[-10000:]
-
-    if all_labels is None and all_predicts is None:
-        all_labels = batch_labels
-        all_predicts = batch_predicts
-    elif all_labels is not None and all_predicts is not None:
-        all_labels = torch.cat((all_labels, batch_labels))
-        all_predicts = torch.cat((all_predicts, batch_predicts))
-
-    conf_matrix = confusion_matrix(all_labels,
-                                   all_predicts,
-                                   labels=list(range(num_classes)))
-
-    probs_matrix = np.zeros(conf_matrix.shape)
-
-    for i in range(probs_matrix.shape[0]):
-        row = conf_matrix[i]
-        if np.sum(row) == 0:
-            probs_row = 0
-        else:
-            probs_row = row / np.sum(row)
-        probs_matrix[i] = probs_row
-
-    probs_matrix = np.around(probs_matrix, decimals=5)
-    return probs_matrix, all_labels, all_predicts
-
-
-# printing the confusion matrix during training
-def print_conf_matrix(confusion_matrix, classes):
-    first_line = "   " + " ".join(
-        ['{:5s}'.format(_class) for _class in classes])
-    print(first_line)
-    for row, _class in zip(confusion_matrix, classes):
-        row_pretty = '{:3s}'.format(_class) + " ".join(
-            ['{:.3f}'.format(number) for number in row])
-        print(row_pretty)
-
-
-# random rotation function
-# credits to Naofumi Tomita
-class Random90Rotation():
-    def __init__(self, degrees=[0, 90, 180, 270]):
-        self.degrees = degrees
-
-    def __call__(self, im):
-        degree = random.sample(self.degrees, k=1)[0]
-        return im.rotate(degree)
-
-
-# instantiate the model
-def create_model(num_layers, pretrain):
-    assert num_layers in [18, 24, 50, 101, 152]
-    architecture = "resnet{}".format(num_layers)
-    model_constructor = getattr(torchvision.models, architecture)
+
+def calculate_confusion_matrix(all_labels: np.ndarray,
```

Comment:
have you tested it for a large number? the reason I had the `10000` stuff is because it took too long to update after having like 100k samples

Reply:
@ntomita @jasonwei20 This commit allocates tensors at the beginning of the training/evaluation loop which store the output.  The confusion matrix is calculated once on the 1d predicted and actual class values at the end of training/evaluation.

Reply:
@jasonwei20 @ntomita We also cannot use the 10000 indexing version since that will return an incorrect confusion matrix when dealing with more than 10000 inputs.